core/app/handler.go

```diff
@@ -0,0 +1,53 @@
+package app
+
+import (
+	"context"
+
+	"google.golang.org/grpc"
+
+	"github.com/cosmos/cosmos-sdk/container"
+)
+
+func NewHandler() *Handler {
+	return &Handler{}
+}
+
+// Handler describes an ABCI app handler.
+type Handler struct {
+	MsgServices   []ServiceImpl
+	QueryServices []ServiceImpl
+
+	// BeginBlocker doesn't take or return any special arguments as this
+	// is the most stable across Tendermint versions and most common need
+	// for modules. Special parameters can be inspected and/or returned
+	// using custom hooks that the app will provide which may vary from
+	// one Tendermint release to another.
+	BeginBlocker func(context.Context) error
+
+	// EndBlocker doesn't take or return any special arguments as this
+	// is the most stable across Tendermint versions and most common need
+	// for modules. Special parameters can be inspected and/or returned
+	// using custom hooks that the app will provide which may vary from
+	// one Tendermint release to another.
+	EndBlocker func(context.Context) error
+}
+
+// RegisterService registers a msg or query service. If the cosmos.msg.v1.service
+// option is set true on the service, then it is registered as a msg service,
+// otherwise it is registered as a query service.
+func (h *Handler) RegisterService(desc *grpc.ServiceDesc, impl interface{}) {
```

Comment:
it's unclear to me where baseapp (more specifically, the router and the middlewares) goes in core

Reply:
I have some ideas but honestly I think I'll need to spike that part out a bit more to have a clear design